### PR TITLE
Introducing typed SequenceNumber[T]

### PR DESCRIPTION
**Pull request by leogrim**
1. SequenceNumber now takes a type parameter. It's no longer a value class because of type erasure conflicts. 
2. @eishay All models, repos and internal routes have been refactored to further enforce this typing: we no longer use Long and since proper bindings are now available, there is no excuse to ever again use Long in service calls instead of SequenceNumber[T](same was already true for Id, please let's avoid using Long for these as well). 
3. As a nice side-effect, typing SequenceNumbers leads to stricter typing of search indices: an index is now typed not only according to the identifier of a document but also by the type of the entities the document contains.
4. @ymatsuda @yingjieMiao I have also refined the typing in IndexerPlugin in order to avoid unsafe casting asInstanceOf[Indexer[...]]. I was able to remove these casts but the extra type parameter I adds some clutter. I'm going to try and improve on this using a type member instead, but this works.
5. @andrewconner Working on this, I found out that sqldb tests have probably been broken for a while (this is true on our Master branch today). So it looks like running kifi-backend's tests does not run them. What do you think?

### CODE REVIEW

File: kifi-backend/modules/search/app/com/keepit/search/graph/user/SearchFriendPlugin.scala [hunk 1]
```diff
@@ -9,17 +9,18 @@ import com.keepit.search.index.IndexerActor
...
-trait SearchFriendGraphPlugin extends IndexerPlugin[SearchFriendIndexer]
+trait SearchFriendGraphPlugin extends IndexerPlugin[SearchFriend, SearchFriendIndexer]
 
 class SearchFriendGraphPluginImpl @Inject()(
   actor: ActorInstance[SearchFriendIndexerActor],
   indexer: SearchFriendIndexer,
   serviceDiscovery: ServiceDiscovery,
   val scheduling: SchedulingProperties
-) extends IndexerPluginImpl[SearchFriendIndexer, SearchFriendIndexerActor](indexer.asInstanceOf[IndexManager[SearchFriendIndexer]], actor, serviceDiscovery) with SearchFriendGraphPlugin
+) extends IndexerPluginImpl[SearchFriend, SearchFriendIndexer, SearchFriendIndexerActor](indexer, actor, serviceDiscovery) with SearchFriendGraphPlugin
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/graph/user/SearchFriendPlugin.scala:21]: @ymatsuda @yingjieMiao 
Got this to compile without a cast

File: kifi-backend/modules/search/app/com/keepit/search/index/Indexer.scala [hunk 2]
```diff
@@ -42,25 +42,25 @@ trait IndexingEventHandler[T] {
...
-  def onFailure(indexable: Indexable[T], e: Throwable): Unit = {
+  def onFailure(indexable: Indexable[T, S], e: Throwable): Unit = {
     _failureCount += 1
     throw e
   }
 }
 
-abstract class Indexer[T](
+abstract class Indexer[T, S, I <: Indexer[T, S, I]](
     indexDirectory: IndexDirectory,
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/index/Indexer.scala:60]: @ymatsuda @yingjieMiao 
To get this to compile without a cast, I had to introduce I here, which represents the type of the subclass down the road. Not sure if I can make it a type member and still get it to compile. Will try.

File: kifi-backend/modules/search/app/com/keepit/search/graph/user/SearchFriendPlugin.scala (hunk 1 above)
Comment (yingjieMiao) [inline kifi-backend/modules/search/app/com/keepit/search/graph/user/SearchFriendPlugin.scala:21]: nice!

**Comment by yingjieMiao** [comment]
lgtm

**Comment by eishay** [comment]
fantastic, good job @leogrim thanks for the initiative!
death to random integers and longs instead of typed id objects!

**Comment by raymondng** [comment]
Nice :+1:

**Pull request merged**